File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/spots/conversationspot/JoinConversationHandler.java
package systems.zlink.samples.supportchat.server.support.spots.conversationspot;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkSpotActorRequestHandler;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class JoinConversationHandler
        implements ZLinkSpotActorRequestHandler<
                ConversationSpot,
                SupportUserActor,
                Messages.JoinConversationReq,
                Messages.JoinConversationRes> {
    @Override
    public CompletionStage<Messages.JoinConversationRes> handle(
            ConversationSpot spot,
            SupportUserActor actor,
            ZLinkMessageContext context,
            Messages.JoinConversationReq request) {
        return CompletableFuture.completedFuture(spot.refreshMembership(actor));
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/conversationspot/handlers/JoinConversationHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotActorRequestHandler
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot.ConversationSpot
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes

class JoinConversationHandler :
    ZLinkSuspendingSpotActorRequestHandler<
        ConversationSpot,
        SupportUserActor,
        JoinConversationReq,
        JoinConversationRes,
    > {
    override suspend fun handle(
        spot: ConversationSpot,
        actor: SupportUserActor,
        context: ZLinkMessageContext,
        request: JoinConversationReq,
    ): JoinConversationRes = spot.refreshMembership(actor)
}
